lib/logstash/bundler.rb

```diff
@@ -0,0 +1,125 @@
+require "bundler"
+require "bundler/cli"
+
+module Bundler
+  # Patch bundler to write a .lock file specific to the version of ruby.
+  # This keeps MRI/JRuby/RBX from conflicting over the Gemfile.lock updates
+  module SharedHelpers
+    def default_lockfile
+      ruby = "#{LogStash::Environment.ruby_engine}-#{LogStash::Environment.ruby_abi_version}"
+      Pathname.new("#{default_gemfile}.#{ruby}.lock")
+    end
+  end
+
+  # Add the Bundler.reset! method which has been added in master but is not in 1.7.9.
+  class << self
+    unless self.method_defined?("reset!")
+      def reset!
+        @definition = nil
+      end
+    end
+
+    def reset_settings
+      # reset!
+      # clear_gemspec_cache
+
+      # remove_instance_variable(:@setup) if defined?(@setup)
+      # remove_instance_variable(:@locked_gems) if defined?(@locked_gems)
+      # remove_instance_variable(:@load) if defined?(@load)
+      remove_instance_variable(:@settings) if defined?(@settings)
+    end
+  end
+end
+
+module LogStash
+  module Bundler
+    # capture any $stdout from the passed block. also trap any exception in that block, in which case the trapped exception will be returned
+    # @param [Proc] the code block to execute
+    # @return [String, Exception] the captured $stdout string and any trapped exception or nil if none
+    def self.capture_stdout
+      begin
+        old_stdout = $stdout
+        $stdout = StringIO.new('', 'w')
+        begin
+          yield
+        rescue => e
+          return [$stdout.string, e]
+        end
+
+        [$stdout.string, nil]
+      ensure
+        $stdout = old_stdout
+      end
+    end
+
+    # execute bundle install and capture any $stdout output. any raised exception in the process will be trapped
+    # and returned. logs errors to $stdout.
+    # @param options [Hash] invoke options with default values, :max_tries => 10, :clean => false, :install => false, :update => false
+    # @param   options[:update] must be either false or a String or an Array of String
+    # @return [String, Exception] the installation captured output and any raised exception or nil if none
+    def self.invoke_bundler!(options = {})
+      options = {:max_tries => 10, :clean => false, :install => false, :update => false, :without => [:development]}.merge(options)
+      options[:without] = Array(options[:without])
+      options[:update] = Array(options[:update]) if options[:update]
+
+      ENV["GEM_PATH"] = LogStash::Environment.logstash_gem_home
+      ENV["BUNDLE_PATH"] = LogStash::Environment.logstash_gem_home
+      ENV["BUNDLE_GEMFILE"] = LogStash::Environment::GEMFILE_PATH
+      ENV["BUNDLE_WITHOUT"] = options[:without].empty? ? "" : options[:without].join(":")
+
+      try = 0
```

Comment:
no particular reason. after refactoring multiple times, I liked this version and had other issues to attend to :P
but I am always open to have neater code :)